components/brave_today/browser/direct_feed_controller.cc

```diff
@@ -86,33 +76,26 @@ mojom::ArticlePtr RustFeedItemToArticle(const FeedItem& rust_feed_item) {
 
 using ParseFeedCallback = base::OnceCallback<void(absl::optional<FeedData>)>;
 void ParseFeedDataOffMainThread(const GURL& feed_url,
-                                const std::string& charset,
                                 std::string body_content,
                                 ParseFeedCallback callback) {
   base::ThreadPool::PostTaskAndReplyWithResult(
       FROM_HERE,
       base::BindOnce(
-          [](const GURL& feed_url, const std::string& charset,
+          [](const GURL& feed_url,
              std::string body_content) -> absl::optional<FeedData> {
```

Comment:
If I understand this PR correctly, this function assumes that the given `body_content` is always in utf-8, doesn't it? Then, can we add a DCHECK for it?

I guess "static bool [Validate](https://source.chromium.org/chromium/chromium/src/+/main:base/i18n/streaming_utf8_validator.h;bpv=1;bpt=1;l=56?gsn=Validate&gs=kythe%3A%2F%2Fchromium.googlesource.com%2Fchromium%2Fsrc%3Flang%3Dc%252B%252B%3Fpath%3Dbase%2Fi18n%2Fstreaming_utf8_validator.h%23_mYw1IExJZkQPyNjiBtdI5ZDU2fR8cU4g4U_wjK9qbw)(const std::[string](https://source.chromium.org/chromium/chromium/src/+/main:buildtools/third_party/libc++/trunk/include/__fwd/string.h;drc=6dc39d60fea45c003424272efdb4c366119a9d7f;bpv=1;bpt=1;l=47)& [string](https://source.chromium.org/chromium/chromium/src/+/main:base/i18n/streaming_utf8_validator.h;bpv=1;bpt=1;l=56?gsn=string&gs=kythe%3A%2F%2Fchromium.googlesource.com%2Fchromium%2Fsrc%3Flang%3Dc%252B%252B%3Fpath%3Dbase%2Fi18n%2Fstreaming_utf8_validator.h%23BXEZzuHAd0TvL7Gc_UXQ6wchLTIxUBG_3I5RQJKAzF4));" could be used but not sure.

Reply:
Not quite - so `body_content` can be in any encoding now - however, the first line of the RSS feed specifies the encoding, and only contains ASCII.

The issue with converting to UTF-8 here is that it would result in a double decode here:
1. Here, when we converted `charset` to UTF-8 (which would stop the Rust string from panicking about non-UTF8 data).
2. In `feed_rs` which would be passed an array of bytes. It would read the encoding header and convert the UTF8 byte stream from `charset` to UTF8 (even though it was already in the right format!).

The fix here is to pass an array of bytes across the FFI instead of a string (which in Rust, must contain UTF8 encoded data). That way we can just let `feed_rs` do the decoding, and we don't have to care about the charset of feeds at all!